libethereum/Block.cpp

```diff
@@ -692,6 +693,29 @@ void Block::performIrregularModifications()
 	}
 }
 
+void Block::updateBlockhashContract()
+{
+	u256 const blockNumber = info().number();
+
+	u256 const metropolisForkBlock = m_sealEngine->chainParams().u256Param("metropolisForkBlock");
+	if (blockNumber == metropolisForkBlock)
+	{
+		m_state.createContract(c_blockhashContractAddress);
+		m_state.setNewCode(c_blockhashContractAddress, bytes(c_blockhashContractCode));
+		m_state.commit(State::CommitBehaviour::KeepEmptyAccounts);
+	}
+
+	if (blockNumber >= metropolisForkBlock)
+	{
+		u256 const nonce = transactionsFrom(SystemAddress);
+		u256 const gas = 1000000;
+		Transaction tx(0, 0, gas, c_blockhashContractAddress, m_previousBlock.hash().asBytes(), nonce);
+		tx.forceSender(SystemAddress);
+		// passing empty lastHashes - assuming the contract doesn't use BLOCKHASH
+		m_state.execute(EnvInfo(info(), {}, 0), *m_sealEngine, tx, Permanence::Committed);
```

Comment:
Performing the call to update the blockhash contract should be similar to CALL from EVM, not to external transaction and should not increment SYSTEM_ADDRESS's nonce.

So the fix is needed here.